File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/device/DeviceControlViewModel.kt
package com.openu60.feature.router.device

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.DeviceParser
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class DeviceControlState(
    val chargeControlLoaded: Boolean = false,
    val powerSaveLoaded: Boolean = false,
    val fastBootLoaded: Boolean = false,
    val autoSleepLoaded: Boolean = false,
    val chargeLimitEnabled: Boolean = false,
    val chargeLimit: Int = 100,
    val hysteresis: Int = 5,
    val powerSave: Boolean = false,
    val fastBoot: Boolean = false,
    val autoSleepEnabled: Boolean = false,
    val autoSleepTimeout: String = "5",
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
    val showRebootConfirm: Boolean = false,
    val showResetConfirm: Boolean = false,
)

@HiltViewModel
class DeviceControlViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(DeviceControlState())
    val state: StateFlow<DeviceControlState> = _state.asStateFlow()

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                // Charge control
                try {
                    val resp = agentClient.getJSON("/api/device/charge-control")
                    @Suppress("UNCHECKED_CAST")
                    val data = resp["data"] as? Map<String, Any?> ?: resp
                    _state.value = _state.value.copy(
                        chargeLimitEnabled = data["charge_limit_enabled"] as? Boolean ?: false,
                        chargeLimit = DeviceParser.asInt(data["charge_limit"]) ?: 100,
                        hysteresis = DeviceParser.asInt(data["hysteresis"]) ?: 5,
                        chargeControlLoaded = true,
                    )
                } catch (_: Exception) {}

                // Power save
                try {
                    val psData = agentClient.getJSON("/api/device/power-save")
                    _state.value = _state.value.copy(
                        powerSave = DeviceParser.asBool(psData["power_saver_mode"]),
                        powerSaveLoaded = true,
                    )
                } catch (_: Exception) {}

                // Fast boot
                try {
                    val fbData = agentClient.getJSON("/api/device/fast-boot")
                    _state.value = _state.value.copy(
                        fastBoot = DeviceParser.asBool(fbData["fast_boot"]),
                        fastBootLoaded = true,
                    )
                } catch (_: Exception) {}

                // Auto sleep
                try {
                    val asData = agentClient.getJSON("/api/device/auto-sleep")
                    @Suppress("UNCHECKED_CAST")
                    val asInner = asData["data"] as? Map<String, Any?> ?: asData
                    _state.value = _state.value.copy(
                        autoSleepEnabled = asInner["enabled"] as? Boolean ?: false,
                        autoSleepTimeout = asInner["timeout"]?.toString() ?: "5",
                        autoSleepLoaded = true,
                    )
                } catch (_: Exception) {}

                _state.value = _state.value.copy(isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun setChargeLimit(enabled: Boolean, limit: Int, hysteresis: Int? = null) {
        val prevEnabled = _state.value.chargeLimitEnabled
        val prevLimit = _state.value.chargeLimit
        val prevHysteresis = _state.value.hysteresis
        _state.value = _state.value.copy(
            chargeLimitEnabled = enabled,
            chargeLimit = limit,
            hysteresis = hysteresis ?: _state.value.hysteresis,
            message = null,
        )
        viewModelScope.launch {
            try {
                val body = mutableMapOf<String, Any>(
                    "charge_limit_enabled" to enabled,
                    "charge_limit" to limit,
                )
                if (hysteresis != null) body["hysteresis"] = hysteresis
                val result = agentClient.putJSON("/api/device/charge-control", body)
                @Suppress("UNCHECKED_CAST")
                val data = result["data"] as? Map<String, Any?>
                val newEnabled = data?.get("charge_limit_enabled") as? Boolean ?: enabled
                val newLimit = DeviceParser.asInt(data?.get("charge_limit")) ?: limit
                val newHysteresis = DeviceParser.asInt(data?.get("hysteresis")) ?: _state.value.hysteresis
                _state.value = _state.value.copy(
                    chargeLimitEnabled = newEnabled,
                    chargeLimit = newLimit,
                    hysteresis = newHysteresis,
                )
            } catch (e: AgentError.Unauthorized) {
                _state.value = _state.value.copy(chargeLimitEnabled = prevEnabled, chargeLimit = prevLimit, hysteresis = prevHysteresis)
                if (authManager.reauthenticate()) setChargeLimit(enabled, limit, hysteresis) else setError(e.message)
            } catch (e: Exception) {
                _state.value = _state.value.copy(chargeLimitEnabled = prevEnabled, chargeLimit = prevLimit, hysteresis = prevHysteresis)
                setError(e.message)
            }
        }
    }

    fun togglePowerSave(enabled: Boolean) {
        val prev = _state.value.powerSave
        _state.value = _state.value.copy(powerSave = enabled, message = null)
        viewModelScope.launch {
            try {
                agentClient.putJSON("/api/device/power-save", mapOf(
                    "deviceInfoList" to mapOf("power_saver_mode" to if (enabled) "1" else "0"),
                ))
            } catch (e: AgentError.Unauthorized) {
                _state.value = _state.value.copy(powerSave = prev)
                if (authManager.reauthenticate()) togglePowerSave(enabled) else setError(e.message)
            } catch (e: Exception) {
                _state.value = _state.value.copy(powerSave = prev)
                setError(e.message)
            }
        }
    }

    fun toggleFastBoot(enabled: Boolean) {
        val prev = _state.value.fastBoot
        _state.value = _state.value.copy(fastBoot = enabled, message = null)
        viewModelScope.launch {
            try {
                agentClient.putJSON("/api/device/fast-boot", mapOf(
                    "fast_boot" to if (enabled) "1" else "0",
                ))
            } catch (e: AgentError.Unauthorized) {
                _state.value = _state.value.copy(fastBoot = prev)
                if (authManager.reauthenticate()) toggleFastBoot(enabled) else setError(e.message)
            } catch (e: Exception) {
                _state.value = _state.value.copy(fastBoot = prev)
                setError(e.message)
            }
        }
    }

    fun setAutoSleep(enabled: Boolean, timeout: String? = null) {
        val prevEnabled = _state.value.autoSleepEnabled
        val prevTimeout = _state.value.autoSleepTimeout
        _state.value = _state.value.copy(
            autoSleepEnabled = enabled,
            autoSleepTimeout = timeout ?: _state.value.autoSleepTimeout,
            message = null,
        )
        viewModelScope.launch {
            try {
                val body = mutableMapOf<String, Any>("enabled" to enabled)
                if (timeout != null) body["timeout"] = timeout
                agentClient.putJSON("/api/device/auto-sleep", body)
            } catch (e: AgentError.Unauthorized) {
                _state.value = _state.value.copy(autoSleepEnabled = prevEnabled, autoSleepTimeout = prevTimeout)
                if (authManager.reauthenticate()) setAutoSleep(enabled, timeout) else setError(e.message)
            } catch (e: Exception) {
                _state.value = _state.value.copy(autoSleepEnabled = prevEnabled, autoSleepTimeout = prevTimeout)
                setError(e.message)
            }
        }
    }

    fun showRebootConfirm() { _state.value = _state.value.copy(showRebootConfirm = true) }
    fun dismissRebootConfirm() { _state.value = _state.value.copy(showRebootConfirm = false) }
    fun showResetConfirm() { _state.value = _state.value.copy(showResetConfirm = true) }
    fun dismissResetConfirm() { _state.value = _state.value.copy(showResetConfirm = false) }

    fun reboot() {
        _state.value = _state.value.copy(showRebootConfirm = false, isLoading = true, message = null)
        viewModelScope.launch {
            try {
                agentClient.postJSON("/api/device/reboot")
                _state.value = _state.value.copy(isLoading = false, message = "Rebooting...", messageIsError = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) reboot() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun factoryReset() {
        _state.value = _state.value.copy(showResetConfirm = false, isLoading = true, message = null)
        viewModelScope.launch {
            try {
                agentClient.postJSON("/api/device/factory-reset")
                _state.value = _state.value.copy(isLoading = false, message = "Factory reset initiated...", messageIsError = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) factoryReset() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}

File: mobile/ios/OpenU60/Features/RouterSettings/Device/DeviceControlViewModel.swift
import SwiftUI

@Observable
@MainActor
final class DeviceControlViewModel {
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false
    var showRebootConfirm: Bool = false
    var showFactoryResetConfirm: Bool = false
    var chargeLimitEnabled: Bool = false
    var chargeLimit: Int = 100
    var hysteresis: Int = 5
    var powerSaveEnabled: Bool = false
    var fastBootEnabled: Bool = false
    var autoSleepEnabled: Bool = false
    var autoSleepTimeout: String = "5"
    private var chargeControlLoaded: Bool = false
    private var powerSaveLoaded: Bool = false
    private var fastBootLoaded: Bool = false
    private var autoSleepLoaded: Bool = false
    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        do {
            let resp = try await client.getJSON("/api/device/charge-control")
            let data = resp["data"] as? [String: Any] ?? resp
            if let enabled = data["charge_limit_enabled"] as? Bool {
                chargeLimitEnabled = enabled
            }
            if let limit = data["charge_limit"] as? Int {
                chargeLimit = limit
            }
            if let hyst = data["hysteresis"] as? Int {
                hysteresis = hyst
            }
            chargeControlLoaded = true
        } catch {
            showMessage("Failed to load charge control status", isError: true)
        }

        do {
            let psData = try await client.postJSON("/api/device/power-save", body: ["deviceInfoList": ["power_saver_mode"]])
            let psMode = psData["power_saver_mode"] as? String ?? ""
            if !psMode.isEmpty {
                let newPowerSave = (psMode == "1")
                if newPowerSave != powerSaveEnabled { powerSaveEnabled = newPowerSave }
            }
            powerSaveLoaded = true
        } catch {
            showMessage("Failed to load power-save settings", isError: true)
        }

        do {
            let fbData = try await client.getJSON("/api/device/fast-boot")
            let data = fbData["data"] as? [String: Any] ?? fbData
            let fbMode = data["fast_boot"] as? String ?? ""
            if !fbMode.isEmpty {
                let newFastBoot = (fbMode == "1")
                if newFastBoot != fastBootEnabled { fastBootEnabled = newFastBoot }
            }
            fastBootLoaded = true
        } catch {
            showMessage("Failed to load fast boot settings", isError: true)
        }

        do {
            let asData = try await client.getJSON("/api/device/auto-sleep")
            let data = asData["data"] as? [String: Any] ?? asData
            if let enabled = data["enabled"] as? Bool {
                autoSleepEnabled = enabled
            }
            if let timeout = data["timeout"] as? String, !timeout.isEmpty {
                autoSleepTimeout = timeout
            }
            autoSleepLoaded = true
        } catch {
            showMessage("Failed to load auto-sleep settings", isError: true)
        }
    }

    func setChargeLimit(enabled: Bool, limit: Int, hysteresis: Int? = nil) async {
        guard chargeControlLoaded else { return }
        let prevEnabled = chargeLimitEnabled
        let prevLimit = chargeLimit
        let prevHysteresis = self.hysteresis
        do {
            var body: [String: Any] = [
                "charge_limit_enabled": enabled,
                "charge_limit": limit,
            ]
            if let hyst = hysteresis {
                body["hysteresis"] = hyst
            }
            let resp = try await client.putJSON("/api/device/charge-control", body: body)
            let data = (resp["data"] as? [String: Any]) ?? resp
            if let newEnabled = data["charge_limit_enabled"] as? Bool {
                chargeLimitEnabled = newEnabled
            }
            if let newLimit = data["charge_limit"] as? Int {
                chargeLimit = newLimit
            }
            if let newHyst = data["hysteresis"] as? Int {
                self.hysteresis = newHyst
            }
        } catch {
            chargeLimitEnabled = prevEnabled
            chargeLimit = prevLimit
            self.hysteresis = prevHysteresis
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
    }

    func setPowerSave(enabled: Bool) async {
        guard powerSaveLoaded else { return }
        do {
            let _ = try await client.putJSON("/api/device/power-save", body: ["deviceInfoList": ["power_saver_mode": enabled ? "1" : "0"]])
        } catch {
            powerSaveEnabled = !enabled
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
    }

    func setFastBoot(enabled: Bool) async {
        guard fastBootLoaded else { return }
        do {
            let _ = try await client.putJSON("/api/device/fast-boot", body: ["fast_boot": enabled ? "1" : "0"])
        } catch {
            fastBootEnabled = !enabled
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
    }

    func setAutoSleep(enabled: Bool, timeout: String? = nil) async {
        guard autoSleepLoaded else { return }
        let prevEnabled = autoSleepEnabled
        let prevTimeout = autoSleepTimeout
        do {
            var body: [String: Any] = ["enabled": enabled]
            if let t = timeout { body["timeout"] = t }
            let resp = try await client.putJSON("/api/device/auto-sleep", body: body)
            let data = resp["data"] as? [String: Any] ?? resp
            if let newEnabled = data["enabled"] as? Bool { autoSleepEnabled = newEnabled }
            if let newTimeout = data["timeout"] as? String { autoSleepTimeout = newTimeout }
        } catch {
            autoSleepEnabled = prevEnabled
            autoSleepTimeout = prevTimeout
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
    }

    func reboot() async {
        isLoading = true

        do {
            let _ = try await client.postJSON("/api/device/reboot")
            showMessage("Router is rebooting...", isError: false)
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func factoryReset() async {
        isLoading = true

        do {
            let _ = try await client.postJSON("/api/device/factory-reset")
            showMessage("Factory reset initiated...", isError: false)
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}
